sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,635 @@
+                                      source.spec.getConnectionConfiguration().getIndex(),
+                                      source.spec.getConnectionConfiguration().getType());
+      Map<String, String> params = new HashMap<>();
+      params.put("scroll", source.spec.getScrollKeepalive());
+      params.put("size", "1");
+      if (source.shardPreference != null) {
+        params.put("preference", "_shards:" + source.shardPreference);
+      }
+      HttpEntity queryEntity = new NStringEntity(query, ContentType.APPLICATION_JSON);
+      response =
+          restClient.performRequest("GET", endPoint, params, queryEntity, new BasicHeader("", ""));
+      JsonObject searchResult = parseResponse(response);
+      scroll = searchResult.getAsJsonPrimitive("_scroll_id").getAsString();
+      return updateCurrent(searchResult);
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      String requestBody = String.format("{\"scroll\" : \"%s\",\"scroll_id\" : \"%s\"}",
+                                        source.spec.getScrollKeepalive(), scroll);
+      HttpEntity scrollEntity = new NStringEntity(requestBody, ContentType.APPLICATION_JSON);
+      Response response = restClient.performRequest("GET", "/_search/scroll",
+                                                    Collections.<String, String>emptyMap(),
+                                                    scrollEntity,
+                                                    new BasicHeader("", ""));
+      JsonObject searchResult = parseResponse(response);
+      return updateCurrent(searchResult);
+    }
+
+    private boolean updateCurrent(JsonObject searchResult) {
+      //stop if no more data
+      if (searchResult.getAsJsonObject("hits").getAsJsonArray("hits").size() == 0) {
+        current = null;
+        return false;
+      }
+      current = searchResult.getAsJsonObject("hits").getAsJsonArray("hits").get(0)
+          .getAsJsonObject().getAsJsonObject("_source").toString();
+      return true;
+    }
+
+    @Override
+    public String getCurrent() throws NoSuchElementException {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
+
+    @Override
+    public void close() throws IOException {
+      // remove the scroll
+      String requestBody = String.format("{\"scroll_id\" : [\"%s\"]}", scroll);
+      HttpEntity entity = new NStringEntity(requestBody,
+                                            ContentType.APPLICATION_JSON);
+      try {
+        restClient.performRequest("DELETE", "/_search/scroll",
+                                  Collections.<String, String>emptyMap(),
+                                  entity, new BasicHeader("", ""));
+      } catch (IOException e) {
+          // Each reader.start() starts a new scroll identified by a scroll_id .ES sometimes gives
+          // the same scroll_id to multiple readers. As a consequence, when these readers call
```

Comment:
This is weird - does it mean closing one reader may make another reader stop working, because its scroll has been deleted and its search context is not being kept alive anymore? That sounds like a very serious bug in Elasticsearch if it is so; is there a public bug entry about it somewhere?